From 25c71c6a7a9f4ec0cccc91fbae277bc0e3db1766 Mon Sep 17 00:00:00 2001
From: Lluis Batlle i Rossell <viric@viric.name>
Date: Thu, 12 Apr 2012 10:24:29 +0200
Subject: [PATCH] Fix crash in the wxActiveXContainer initialisation.

Using activex created based on the MSVC Template for MFC ActiveX, the
code crashed due to a lack of a hWnd.

I moved the code around, so we enable the activex before doing window
operations with it. For the tests I've done, the wxWebViewIE and my MFC activex
all work fine after this change, although I don't understand the full details
of that initialisation, and I can't really forecast if I could have broken
anything.

Based on
http://stackoverflow.com/questions/6055216/how-can-i-programmatically-activate-an-ole-control-via-com-interfaces
---
 src/msw/ole/activex.cpp |   22 +++++++++++-----------
 1 files changed, 11 insertions(+), 11 deletions(-)

diff --git a/src/msw/ole/activex.cpp b/src/msw/ole/activex.cpp
index d161e6d..6b1b3d1 100644
--- a/src/msw/ole/activex.cpp
+++ b/src/msw/ole/activex.cpp
@@ -1105,16 +1105,20 @@ void wxActiveXContainer::CreateActiveX(REFIID iid, IUnknown* pUnk)
 
     m_oleObjectHWND = 0;
 
-    if (m_oleInPlaceObject.IsOk())
-    {
-        hret = m_oleInPlaceObject->GetWindow(&m_oleObjectHWND);
-        if (SUCCEEDED(hret))
-            ::SetActiveWindow(m_oleObjectHWND);
-    }
-
 
     if (! (dwMiscStatus & OLEMISC_INVISIBLEATRUNTIME))
     {
+        hret = m_oleObject->DoVerb(OLEIVERB_INPLACEACTIVATE, NULL,
+            m_clientSite, 0, (HWND)m_realparent->GetHWND(), NULL);
+        CHECK_HR(hret);
+
+        if (m_oleInPlaceObject.IsOk())
+        {
+            hret = m_oleInPlaceObject->GetWindow(&m_oleObjectHWND);
+            CHECK_HR(hret);
+            ::SetActiveWindow(m_oleObjectHWND);
+        }
+
         RECT posRect;
         wxCopyRectToRECT(m_realparent->GetClientSize(), posRect);
 
@@ -1124,10 +1128,6 @@ void wxActiveXContainer::CreateActiveX(REFIID iid, IUnknown* pUnk)
             m_oleInPlaceObject->SetObjectRects(&posRect, &posRect);
         }
 
-        hret = m_oleObject->DoVerb(OLEIVERB_INPLACEACTIVATE, NULL,
-            m_clientSite, 0, (HWND)m_realparent->GetHWND(), &posRect);
-        CHECK_HR(hret);
-
         hret = m_oleObject->DoVerb(OLEIVERB_SHOW, 0, m_clientSite, 0,
             (HWND)m_realparent->GetHWND(), &posRect);
         CHECK_HR(hret);
-- 
1.7.5.1

